pkg/webui/components/map/index.js

```diff
@@ -34,74 +36,143 @@ Leaflet.Icon.Default.mergeOptions({
+          longitude: PropTypes.number,
+          latitude: PropTypes.number,
+        }),
       }),
-    ).isRequired,
+    ),
+    onClick: PropTypes.func,
     // Widget is a boolean used to add a class name to the map container div for styling.
     widget: PropTypes.bool,
   }
 
   static defaultProps = {
     leafletConfig: {},
+    className: undefined,
     widget: false,
+    markers: [],
+    onClick: () => null,
+    mapCenter: undefined,
+    clickable: false,
   }
 
-  getMapCenter(markers) {
-    // This will calculate zoom and map center long/lang based on all markers provided.
-    // Currently it just returns the first marker.
-    // TODO: action (https://github.com/TheThingsNetwork/lorawan-stack/issues/1241)
-    return markers[0]
-  }
+  constructor(props) {
+    super(props)
 
-  createMap(config, id) {
-    this.map = Leaflet.map(id, {
-      ...config,
-    })
+    this.state = {
+      zoomLevel: props.leafletConfig.zoom,
+      mapCenter: defaultLocation,
+      loading: props.markers.length === 0,
+    }
   }
 
-  createMarkers(markers) {
-    markers.map(marker =>
-      Leaflet.marker([marker.position.latitude, marker.position.longitude]).addTo(this.map),
-    )
+  static getDerivedStateFromProps(props, state) {
+    const { mapCenter } = props
+    if ('mapCenter' in props && mapCenter !== undefined) {
+      return { mapCenter }
+    }
+    return null
   }
 
+  // fix the issue where tiles sometimes partially load
+  @bind
   componentDidMount() {
-    const { id, markers } = this.props
+    const { markers } = this.props
 
-    const { position } = markers.length >= 1 ? this.getMapCenter(markers) : markers[0]
+    const map = this.refs.map.leafletElement
+    setTimeout(() => {
+      map.invalidateSize()
+    }, 250)
 
-    const config = {
-      layers: [
-        Leaflet.tileLayer('http://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png', {
-          attribution: '&copy; <a href="http://osm.org/copyright">OpenStreetMap</a> contributors',
-        }),
-      ],
-      center: [position.latitude, position.longitude],
-      zoom: 11,
-      minZoom: 1,
-      ...this.props.leafletConfig,
+    if (markers.length === 0) {
+      if ('geolocation' in navigator) {
+        navigator.geolocation.getCurrentPosition(
+          position => {
+            this.setState({
+              mapCenter: [position.coords.latitude, position.coords.longitude],
+              loading: false,
+            })
+          },
+          error => {
+            this.setState({ mapCenter: defaultLocation, loading: false })
+          },
+        )
+      } else {
+        this.setState({ mapCenter: defaultLocation, loading: false })
+      }
     }
+  }
 
-    this.createMap(config, id)
-    this.createMarkers(markers, id)
+  @bind
+  handleClick(event) {
+    const { onClick, clickable } = this.props
+    if (!clickable) {
+      return
+    }
+    const map = this.refs.map.leafletElement
+    this.setState({ mapCenter: map.getCenter() })
```

Comment:
I think the map component itself should stay neutral. If we make too many assumptions about the map behavior we will have a hard time using the map for other purposes too. You can already see that as your implementation now needs the `clickable` prop. The map should be first and foremost concerned with rendering a map based on its props.

So instead of keeping track of the map center in the state in the map component itself, that should be done by a parent component (in this case the `<LocationForm />`)